File: server/services/log-parser.ts
interface LogSegment {
  text: string;
  timestamp?: string;
  level?: string;
  component?: string;
  message: string;
}

class LogParser {
  // Common telecom log patterns
  private patterns = {
    standard: /^(\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+\[([^\]]+)\]\s+\[([^\]]+)\]\s+-\s+(.+)$/,
    cisco: /^(\w+\s+\d+\s+\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+(\w+):\s+%([^:]+):\s+(.+)$/,
    nokia: /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?Z?)\s+(\w+)\s+(\w+)\s+(.+)$/,
    huawei: /^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\+\d{2}:\d{2})?)\s+(\w+)\s+(\S+)\s+(.+)$/,
    ericsson: /^(\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+(\w+)\s+(\w+)\s+(.+)$/
  };

  parseLog(content: string): LogSegment[] {
    // Split the log content into lines
    const lines = content.split(/\r?\n/).filter(line => line.trim() !== '');
    
    return lines.map(line => this.parseLine(line));
  }

  private parseLine(line: string): LogSegment {
    // Try to match the line against each pattern
    for (const [type, pattern] of Object.entries(this.patterns)) {
      const match = line.match(pattern);
      if (match) {
        return {
          text: line,
          timestamp: match[1],
          level: match[2],
          component: match[3],
          message: match[4]
        };
      }
    }
    
    // If no patterns match, just return the line as is
    return {
      text: line,
      message: line
    };
  }

  // Split long logs into manageable chunks for embedding
  segmentLogForEmbedding(content: string, maxChunkSize: number = 512): string[] {
    const segments: string[] = [];
    const lines = content.split(/\r?\n/).filter(line => line.trim() !== '');
    
    let currentChunk = '';
    
    for (const line of lines) {
      if (currentChunk.length + line.length > maxChunkSize && currentChunk.length > 0) {
        segments.push(currentChunk);
        currentChunk = '';
      }
      
      currentChunk += (currentChunk.length > 0 ? '\n' : '') + line;
    }
    
    if (currentChunk.length > 0) {
      segments.push(currentChunk);
    }
    
    return segments;
  }

  // Check if content appears to be a valid telecom log
  isValidTelecomLog(content: string): boolean {
    const lines = content.split(/\r?\n/).filter(line => line.trim() !== '');
    
    // Check at least a few lines to determine if this looks like a telecom log
    let matchedLines = 0;
    const linesToCheck = Math.min(lines.length, 5);
    
    for (let i = 0; i < linesToCheck; i++) {
      for (const pattern of Object.values(this.patterns)) {
        if (lines[i].match(pattern)) {
          matchedLines++;
          break;
        }
      }
    }
    
    // If more than half of the checked lines match our patterns, consider it valid
    return matchedLines >= Math.ceil(linesToCheck / 2);
  }
  
  // Extract timestamps from a log for timeline analysis
  extractTimestamps(content: string): { timestamp: string, line: string }[] {
    const result: { timestamp: string, line: string }[] = [];
    const lines = content.split(/\r?\n/).filter(line => line.trim() !== '');
    
    for (const line of lines) {
      // Try to extract timestamp using the patterns
      for (const pattern of Object.values(this.patterns)) {
        const match = line.match(pattern);
        if (match && match[1]) {
          result.push({
            timestamp: match[1],
            line
          });
          break;
        }
      }
    }
    
    return result;
  }
}

export const logParser = new LogParser();

File: python_backend/services/log_parser.py
import re
from typing import List, Dict, Any
from datetime import datetime

class LogParser:
    """Parser for telecom log files"""
    
    def __init__(self):
        """Initialize with regex patterns for different log formats"""
        # Pattern for standard log format with timestamp, level, component, and message
        self.standard_pattern = re.compile(
            r'(\d{4}-\d{2}-\d{2}[T ]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?)\s+'
            r'(?:([A-Z]+)\s+)?'  # Optional log level (INFO, ERROR, etc.)
            r'(?:\[([^\]]+)\]\s+)?'  # Optional component in square brackets
            r'(.*)'  # Message
        )
        
        # Pattern for Cisco-style logs
        self.cisco_pattern = re.compile(
            r'(\w+\s+\d+\s+\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+'
            r'(?:([A-Z0-9-]+):\s+)?'  # Facility/severity
            r'(?:%([A-Z0-9-]+)(?:-\d+)?:\s+)?'  # Message code
            r'(.*)'  # Message
        )
        
        # Pattern for Nokia/Alcatel-Lucent logs
        self.nokia_pattern = re.compile(
            r'(\d{4}/\d{2}/\d{2}\s+\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+'
            r'(?:([A-Z]+)\s+)?'  # Optional log level
            r'(?:\[(\w+(?:\-\w+)*)\]\s+)?'  # Optional module
            r'(.*)'  # Message
        )
        
        # Pattern for Huawei logs
        self.huawei_pattern = re.compile(
            r'(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+'
            r'(?:([A-Z]+)\s+)?'  # Optional severity
            r'(?:\[([^\]]+)\]\s+)?'  # Optional product/module
            r'(.*)'  # Message
        )
        
        # Pattern for Ericsson logs
        self.ericsson_pattern = re.compile(
            r'(\d{4}-\d{2}-\d{2}\s+\d{2}:\d{2}:\d{2}(?:\.\d+)?)\s+'
            r'(?:(\w+)\s+)?'  # Optional log level
            r'(?:(\w+(?:\.\w+)*)\s+)?'  # Optional component
            r'(.*)'  # Message
        )
        
        # Combined pattern to check if it's a telecom log
        self.telecom_log_pattern = re.compile(
            r'(\d{4}[-/]\d{2}[-/]\d{2}[T ]\d{2}:\d{2}:\d{2}|'
            r'\w+\s+\d+\s+\d{2}:\d{2}:\d{2})'
        )
    
    def parse_log(self, content: str) -> List[Dict[str, Any]]:
        """Parse log content into segments with metadata"""
        lines = content.splitlines()
        segments = []
        
        for line in lines:
            if not line.strip():
                continue
            
            parsed = self.parse_line(line)
            if parsed:
                segments.append(parsed)
            else:
                # If line doesn't match any pattern, add it as raw text
                segments.append({
                    "text": line,
                    "message": line
                })
        
        return segments
    
    def parse_line(self, line: str) -> Dict[str, Any]:
        """Parse a single log line"""
        # Try each pattern
        patterns = [
            self.standard_pattern,
            self.cisco_pattern,
            self.nokia_pattern,
            self.huawei_pattern,
            self.ericsson_pattern
        ]
        
        for pattern in patterns:
            match = pattern.match(line)
            if match:
                timestamp, level, component, message = match.groups()
                
                return {
                    "text": line,
                    "timestamp": timestamp,
                    "level": level,
                    "component": component,
                    "message": message
                }
        
        return None
    
    def segment_log_for_embedding(self, content: str, max_chunk_size: int = 512) -> List[str]:
        """Segment log content into reasonable chunks for embedding"""
        lines = content.splitlines()
        
        # Group lines into segments
        segments = []
        current_segment = []
        current_length = 0
        
        for line in lines:
            line_length = len(line)
            
            # If adding this line would exceed max_chunk_size, start a new segment
            if current_length + line_length > max_chunk_size and current_segment:
                segments.append("\n".join(current_segment))
                current_segment = []
                current_length = 0
            
            # Add line to current segment
            current_segment.append(line)
            current_length += line_length
        
        # Add the last segment if it's not empty
        if current_segment:
            segments.append("\n".join(current_segment))
        
        return segments
    
    def is_valid_telecom_log(self, content: str) -> bool:
        """Check if the content looks like a valid telecom log"""
        # Take a sample of the first 20 non-empty lines
        lines = [line for line in content.splitlines() if line.strip()][:20]
        
        # Count how many lines match our telecom log pattern
        matches = 0
        for line in lines:
            if self.telecom_log_pattern.search(line):
                matches += 1
        
        # If at least 50% of the lines match, consider it a valid telecom log
        return len(lines) > 0 and matches / len(lines) >= 0.5
    
    def extract_timestamps(self, content: str) -> List[Dict[str, str]]:
        """Extract timestamps from log content with their associated lines"""
        lines = content.splitlines()
        timestamps = []
        
        for line in lines:
            for pattern in [
                self.standard_pattern,
                self.cisco_pattern,
                self.nokia_pattern,
                self.huawei_pattern,
                self.ericsson_pattern
            ]:
                match = pattern.match(line)
                if match and match.group(1):  # If we have a timestamp
                    timestamps.append({
                        "timestamp": match.group(1),
                        "line": line
                    })
                    break
        
        return timestamps